File: src/main/java/me/brendn/oak/mcbeta/mixin/world/MixinWorld.java
package me.brendn.oak.mcbeta.mixin.world;

import com.flowpowered.math.vector.Vector3i;
import com.google.common.collect.Lists;
import me.brendn.oak.api.common.entity.Entity;
import me.brendn.oak.api.common.world.World;
import net.minecraft.src.WorldInfo;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;

import java.util.Collection;
import java.util.List;

@Mixin(net.minecraft.src.World.class)
public abstract class MixinWorld implements World {

	@Shadow public abstract int getBlockId(int i, int j, int k);

	@Shadow public abstract int getBlockMetadata(int i, int j, int k);

	@Shadow public abstract boolean isBlockNormalCube(int i, int j, int k);

	@Shadow public abstract boolean isBlockOpaqueCube(int i, int j, int k);

	@Shadow public abstract float getBrightness(int i, int j, int k, int l);

	@Shadow public abstract float getLightBrightness(int i, int j, int k);

	@Shadow public abstract long getRandomSeed();

	@Shadow public abstract WorldInfo getWorldInfo();

	@Shadow public List<net.minecraft.src.Entity> loadedEntityList;

	@Override
	public int getBlockID(Vector3i pos) {
		return getBlockId(pos.getX(), pos.getY(), pos.getZ());
	}

	@Override
	public int getBlockMetadata(Vector3i pos) {
		return getBlockMetadata(pos.getX(), pos.getY(), pos.getZ());
	}

	@Override
	public boolean isBlockOpaque(Vector3i pos) {
		return isBlockOpaqueCube(pos.getX(), pos.getY(), pos.getZ());
	}

	@Override
	public boolean isBlockNormal(Vector3i pos) {
		return isBlockNormalCube(pos.getX(), pos.getY(), pos.getZ());
	}

	@Override
	public float getBrightness(Vector3i pos, int lightValue) {
		return getBrightness(pos.getX(), pos.getY(), pos.getZ(), lightValue);
	}

	@Override
	public float getLightBrightness(Vector3i pos) {
		return getLightBrightness(pos.getX(), pos.getY(), pos.getZ());
	}

	@Override
	public long getSeed() {
		return getRandomSeed();
	}

	@Override
	public boolean isRaining() {
		return getWorldInfo().getRaining();
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<Entity> getLoadedEntities() {
		return Lists.newArrayList((Collection<Entity>) (Object) this.loadedEntityList);
	}
}

File: src/main/kotlin/me/brendn/oak/mcbeta/mixin/world/MixinWorld.kt
package me.brendn.oak.mcbeta.mixin.world

import com.flowpowered.math.vector.Vector3i
import com.google.common.collect.ImmutableList
import com.google.common.collect.Lists
import me.brendn.oak.api.common.entity.Entity
import me.brendn.oak.api.common.world.World
import net.minecraft.src.WorldInfo
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow

@Mixin(net.minecraft.src.World::class)
abstract class MixinWorld : World {

	@Shadow abstract fun getBlockId(i: Int, j: Int, k: Int): Int

	@Shadow abstract fun getBlockMetadata(i: Int, j: Int, k: Int): Int

	@Shadow abstract fun isBlockNormalCube(i: Int, j: Int, k: Int): Boolean

	@Shadow abstract fun isBlockOpaqueCube(i: Int, j: Int, k: Int): Boolean

	@Shadow abstract fun getBrightness(i: Int, j: Int, k: Int, l: Int): Float

	@Shadow abstract fun getLightBrightness(i: Int, j: Int, k: Int): Float

	@get:Shadow abstract val randomSeed: Long

	@get:Shadow abstract val worldInfo: WorldInfo

	@Shadow var loadedEntityList: List<net.minecraft.src.Entity>? = null

	override fun getBlockID(pos: Vector3i): Int {
		return getBlockId(pos.x, pos.y, pos.z)
	}

	override fun getBlockMetadata(pos: Vector3i): Int {
		return getBlockMetadata(pos.x, pos.y, pos.z)
	}

	override fun isBlockOpaque(pos: Vector3i): Boolean {
		return isBlockOpaqueCube(pos.x, pos.y, pos.z)
	}

	override fun isBlockNormal(pos: Vector3i): Boolean {
		return isBlockNormalCube(pos.x, pos.y, pos.z)
	}

	override fun getBrightness(pos: Vector3i, lightValue: Int): Float {
		return getBrightness(pos.x, pos.y, pos.z, lightValue)
	}

	override fun getLightBrightness(pos: Vector3i): Float {
		return getLightBrightness(pos.x, pos.y, pos.z)
	}

	override fun getSeed(): Long {
		return randomSeed
	}

	override fun isRaining(): Boolean {
		return worldInfo.raining
	}

	@Suppress("UNCHECKED_CAST")
	override fun getLoadedEntities(): List<Entity> {
		return ImmutableList.copyOf((this.loadedEntityList as Collection<Entity>?)!!)
	}
}